### PR TITLE
add select primary calendar

### PR DESCRIPTION
**Pull request by KATT**
## What does this PR do?

- Allows the user to select a destination calendar for bookings.-

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How should this be tested?

- [x] On the integrations page, select a destination calendar and make a booking. The new booking should be created only on the selected destination.
- [x] On the onboarding, the selector for destination calendar should be displayed
- [x] Test using Apple Calendars
- [x] Test using Google Calendars
- [ ] Test using CalDav Calendars
- [ ] Test using Office 365 Calendars

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code and corrected any misspellings
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by zomars** [comment]
Moved my comment here from #1155 

> NOTE: we shouldn't forget to add the primary calendar selector in the onboarding: 
>
> Sin título

File: prisma/schema.prisma [hunk 1]
```diff
@@ -70,6 +70,14 @@ enum UserPlan {
   PRO
 }
 
+model UserDestinationCalendar {
+  id          Int    @id
+  integration String
+  externalId  String
+
+  user User @relation(fields: [id], references: [id])
+}
```
Comment (zomars) [inline prisma/schema.prisma:79]: Hey @KATT do you know what's the main difference between `UserDestinationCalendar` and `BookingDestinationCalendar` ?

File: ee/pages/api/integrations/stripepayment/webhook.ts [hunk 2]
```diff
@@ -91,7 +92,7 @@ async function handlePaymentSuccess(event: Stripe.Event) {
   if (booking.location) evt.location = booking.location;
 
   if (booking.confirmed) {
-    const eventManager = new EventManager(user.credentials);
```
Comment (zomars) [inline ee/pages/api/integrations/stripepayment/webhook.ts:94]: Fixed a regression introduced with the `EventManager` rework.

File: lib/CalEventParser.ts [hunk 3]
```diff
@@ -68,7 +68,7 @@ export const getLocation = (calEvent: CalendarEvent) => {
     return calEvent.additionInformation.hangoutLink;
   }
 
-  return providerName || calEvent.location;
+  return providerName || calEvent.location || "";
```
Comment (zomars) [inline lib/CalEventParser.ts:71]: Type fixes

File: lib/calendarClient.ts [hunk 4]
```diff
@@ -98,72 +99,32 @@ function getCalendarAdapterOrNull(credential: Credential): CalendarApiAdapter |
...
-          return Office365CalendarApiAdapter(cred);
-        case "caldav_calendar":
-          return new CalDavCalendar(cred);
-        case "apple_calendar":
-          return new AppleCalendar(cred);
-        default:
-          return; // unknown credential, could be legacy? In any case, ignore
-      }
-    })
-    .flatMap((item) => (item ? [item as CalendarApiAdapter] : []));
```
Comment (zomars) [inline lib/calendarClient.ts:120]: Removed deprecated code

File: lib/calendarClient.ts [hunk 5]
```diff
@@ -98,72 +99,32 @@ function getCalendarAdapterOrNull(credential: Credential): CalendarApiAdapter |
...
-  });
-
-/**
- *
- * @param withCredentials
- * @deprecated
- */
-const listCalendars = (withCredentials: Credential[]) =>
-  Promise.all(calendars(withCredentials).map((c) => c.listCalendars())).then((results) =>
-    results.reduce((acc, calendars) => acc.concat(calendars), []).filter((c) => c != null)
```
Comment (zomars) [inline lib/calendarClient.ts:141]: More deprecated code removal

File: lib/calendarClient.ts [hunk 6]
```diff
@@ -98,72 +99,32 @@ function getCalendarAdapterOrNull(credential: Credential): CalendarApiAdapter |
...
 
-  if (!creationResult) {
-    return {
-      type: credential.type,
-      success,
-      uid,
-      originalEvent: calEvent,
-    };
-  }
-
```
Comment (zomars) [inline lib/calendarClient.ts:166]: Redundant return

File: lib/calendarClient.ts [hunk 7]
```diff
@@ -179,28 +140,18 @@ const updateEvent = async (
...
 
-  if (!updatedResult) {
-    return {
-      type: credential.type,
-      success,
-      uid,
-      originalEvent: calEvent,
-    };
-  }
-
```
Comment (zomars) [inline lib/calendarClient.ts:203]: Redundant return

File: lib/integrations/Apple/AppleCalendarAdapter.ts [hunk 8]
```diff
@@ -1,346 +1,10 @@
...
-
-export class AppleCalendar implements CalendarApiAdapter {
-  private url: string;
-  private credentials: Record<string, string>;
-  private headers: Record<string, string>;
-  private readonly integrationName: string = "apple_calendar";
+import { BaseCalendarApiAdapter } from "@lib/BaseCalendarApiAdapter";
+import { CalendarApiAdapter } from "@lib/calendarClient";
 
+export class AppleCalendar extends BaseCalendarApiAdapter implements CalendarApiAdapter {
```
Comment (zomars) [inline lib/integrations/Apple/AppleCalendarAdapter.ts:6]: Abstracted to `BaseCalendarApiAdapter`

File: lib/integrations/CalDav/CalDavCalendarAdapter.ts [hunk 9]
```diff
@@ -1,347 +1,10 @@
...
-
-export class CalDavCalendar implements CalendarApiAdapter {
-  private url: string;
-  private credentials: Record<string, string>;
-  private headers: Record<string, string>;
-  private readonly integrationName: string = "caldav_calendar";
+import { BaseCalendarApiAdapter } from "@lib/BaseCalendarApiAdapter";
+import { CalendarApiAdapter } from "@lib/calendarClient";
 
+export class CalDavCalendar extends BaseCalendarApiAdapter implements CalendarApiAdapter {
```
Comment (zomars) [inline lib/integrations/CalDav/CalDavCalendarAdapter.ts:6]: Abstracted to `BaseCalendarApiAdapter`

File: lib/integrations/GoogleCalendar/GoogleCalendarApiAdapter.ts [hunk 10]
```diff
@@ -10,18 +10,30 @@ export interface ConferenceData {
...
+  }
+
+  isTokenExpiring() {
+    return super.isTokenExpiring();
+  }
+
+  async refreshToken(token: string | null | undefined) {
+    return super.refreshToken(token);
+  }
+}
```
Comment (zomars) [inline lib/integrations/GoogleCalendar/GoogleCalendarApiAdapter.ts:25]: So TS stops complaining.

File: prisma/schema.prisma [hunk 11]
```diff
@@ -70,39 +71,53 @@ enum UserPlan {
   PRO
 }
 
+model DestinationCalendar {
+  id          Int        @id @default(autoincrement())
+  integration String
+  externalId  String
+  user        User?      @relation(fields: [userId], references: [id])
+  userId      Int?       @unique
```
Comment (zomars) [inline prisma/schema.prisma:79]: For now we only check for the user destination calendar, but...

File: prisma/schema.prisma [hunk 12]
```diff
@@ -70,39 +71,53 @@ enum UserPlan {
...
+  booking     Booking?   @relation(fields: [bookingId], references: [id])
+  bookingId   Int?       @unique
```
Comment (zomars) [inline prisma/schema.prisma:81]: We may use this for historical purposes, cancelling, etc.

File: prisma/schema.prisma [hunk 13]
```diff
@@ -70,39 +71,53 @@ enum UserPlan {
...
+  eventType   EventType? @relation(fields: [eventTypeId], references: [id])
+  eventTypeId Int?       @unique
```
Comment (zomars) [inline prisma/schema.prisma:83]: We may override the user destination calendar and choose per event type in the future.

File: lib/notEmpty.ts [hunk 14]
```diff
@@ -0,0 +1,3 @@
+const notEmpty = <T>(value: T): value is NonNullable<typeof value> => !!value;
+
+export default notEmpty;
```
Comment (zomars) [inline lib/notEmpty.ts:3]: So we can filter out items in a type-safe manner.

**Review by zomars — APPROVED** [review]
Ready for review

**Comment by PeerRich** [comment]
image
migration missing on staging?

**Pull request merged**